R/verb-pivot-longer.R

```diff
@@ -0,0 +1,235 @@
+                                  names_ptypes = list(),
+                                  names_transform = list(),
+                                  names_repair = "check_unique",
+                                  values_to = "value",
+                                  values_drop_na = FALSE,
+                                  values_ptypes,
+                                  values_transform = list(),
+                                  ...) {
+  if (!is_missing(values_ptypes)) {
+    abort("The `values_ptypes` argument is not supported for remote back-ends")
+  }
+
+  ellipsis::check_dots_empty()
+
+  cols <- enquo(cols)
+
+  spec <- tidyr::build_longer_spec(simulate_vars(data), !!cols,
+    names_to = names_to,
+    values_to = values_to,
+    names_prefix = names_prefix,
+    names_sep = names_sep,
+    names_pattern = names_pattern,
+    names_ptypes = names_ptypes,
+    names_transform = names_transform
+  )
+
+  dbplyr_pivot_longer_spec(data, spec,
+    names_repair = names_repair,
+    values_drop_na = values_drop_na,
+    values_transform = values_transform
+  )
+}
+
+dbplyr_pivot_longer_spec <- function(data,
+                                     spec,
+                                     names_repair = "check_unique",
+                                     values_drop_na = FALSE,
+                                     values_transform = list()) {
+  spec <- check_spec(spec)
+  # .seq col needed if different input columns are mapped to the same output
+  # column
+  spec <- deduplicate_spec(spec, data)
+
+  id_cols <- syms(setdiff(colnames(data), spec$.name))
+
+  spec_split <- vctrs::vec_split(spec, spec[, -(1:2)])
+  data_long_list <- purrr::map(
+    vctrs::vec_seq_along(spec_split),
+    function(idx) {
+      row <- spec_split$val[[idx]][, 1:2]
+      keys <- spec_split$key[idx, ]
+      keys$.seq <- NULL
+
+      measure_cols_exprs <- get_measure_column_exprs(
+        row[[".name"]],
+        row[[".value"]],
+        values_transform
+      )
+
+      transmute(
+        data,
+        !!!id_cols,
+        !!!keys,
+        !!!measure_cols_exprs
+      )
+    }
+  )
+
+  data_long <- purrr::reduce(data_long_list, union_all)
+
+
+  if (values_drop_na) {
+    value_cols <- unique(spec$.value)
+
+    data_long <- dplyr::filter_at(
+      data_long,
+      value_cols,
+      dplyr::all_vars(!is.na(.))
+    )
+  }
+
+  data_long
+}
+
+get_measure_column_exprs <- function(name, value, values_transform) {
+  # idea copied from `partial_eval_across`
+  measure_funs <- syms(purrr::map_chr(values_transform, find_fun))
+
+  measure_cols <- set_names(syms(name), value)
+  purrr::imap(
+    measure_cols,
+    ~ {
+      f_trans <- measure_funs[[.y]]
+
+      if (is_null(f_trans)) {
+        .x
+      } else {
+        expr((!!f_trans)(!!.x))
+      }
+    }
+  )
+}
+
+# The following is copy-pasted from tidyr
+# with the exception of `simplifyPieces()` which I have quickly hacked in R
```

Comment:
Can you add a comment pointing to https://github.com/tidyverse/tidyr/issues/1087 please?

Reply:
And use exclusion comments to make sure that it's not counted by test coverage: https://covr.r-lib.org/#exclusion-comments